From 29ef004ee234b476bb25a286e5d05b3b976f0e74 Mon Sep 17 00:00:00 2001
From: Jernej Skrabec <jernej.skrabec@gmail.com>
Date: Thu, 1 Sep 2022 17:36:53 +0200
Subject: [PATCH] ASoC: AC200: Initial driver

Add an ASoC codec driver for the analog audio block in the
X-Powers AC200 mixed-signal companion IC. Support its I2S interface,
playback and capture mixers, microphone inputs, and line output.

Signed-off-by: Jernej Skrabec <jernej.skrabec@gmail.com>
---
 sound/soc/codecs/Kconfig  |  13 +
 sound/soc/codecs/Makefile |   2 +
 sound/soc/codecs/ac200.c  | 958 ++++++++++++++++++++++++++++++++++++++
 3 files changed, 973 insertions(+)
 create mode 100644 sound/soc/codecs/ac200.c

diff --git a/sound/soc/codecs/Kconfig b/sound/soc/codecs/Kconfig
index 252f683be3c1..04c374f10a1a 100644
--- a/sound/soc/codecs/Kconfig
+++ b/sound/soc/codecs/Kconfig
@@ -16,6 +16,7 @@ config SND_SOC_ALL_CODECS
 	depends on COMPILE_TEST
 	imply SND_SOC_88PM860X
 	imply SND_SOC_AB8500_CODEC
+	imply SND_SOC_AC200_CODEC
 	imply SND_SOC_AC97_CODEC
 	imply SND_SOC_AD1836
 	imply SND_SOC_AD193X_SPI
@@ -422,6 +423,18 @@ config SND_SOC_AB8500_CODEC
 	tristate
 	depends on ABX500_CORE
 
+config SND_SOC_AC200_CODEC
+	tristate "AC200 Codec"
+	depends on MFD_AC200
+	help
+	  Enable support for the analog audio codec in the X-Powers AC200
+	  mixed-signal companion IC. The codec supports stereo playback and
+	  capture through its I2S interface, along with microphone and line
+	  output paths.
+
+	  To compile this driver as a module, choose M here: the module
+	  will be called snd-soc-ac200.
+
 config SND_SOC_AC97_CODEC
 	tristate "Build generic ASoC AC97 CODEC driver"
 	select SND_AC97_CODEC
diff --git a/sound/soc/codecs/Makefile b/sound/soc/codecs/Makefile
index aa0396e5b575..97b119728e98 100644
--- a/sound/soc/codecs/Makefile
+++ b/sound/soc/codecs/Makefile
@@ -1,6 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 snd-soc-88pm860x-y := 88pm860x-codec.o
 snd-soc-ab8500-codec-y := ab8500-codec.o
+snd-soc-ac200-y := ac200.o
 snd-soc-ac97-y := ac97.o
 snd-soc-ad1836-y := ad1836.o
 snd-soc-ad193x-y := ad193x.o
@@ -441,6 +442,7 @@ snd-soc-simple-mux-y := simple-mux.o
 
 obj-$(CONFIG_SND_SOC_88PM860X)	+= snd-soc-88pm860x.o
 obj-$(CONFIG_SND_SOC_AB8500_CODEC)	+= snd-soc-ab8500-codec.o
+obj-$(CONFIG_SND_SOC_AC200_CODEC)	+= snd-soc-ac200.o
 obj-$(CONFIG_SND_SOC_AC97_CODEC)	+= snd-soc-ac97.o
 obj-$(CONFIG_SND_SOC_AD1836)	+= snd-soc-ad1836.o
 obj-$(CONFIG_SND_SOC_AD193X)	+= snd-soc-ad193x.o
diff --git a/sound/soc/codecs/ac200.c b/sound/soc/codecs/ac200.c
new file mode 100644
index 000000000000..3a029fbc101e
--- /dev/null
+++ b/sound/soc/codecs/ac200.c
@@ -0,0 +1,958 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+/*
+ * X-Powers AC200 Codec Driver
+ *
+ * Copyright (C) 2022 Jernej Skrabec <jernej.skrabec@gmail.com>
+ */
+
+#include <linux/module.h>
+#include <linux/regmap.h>
+#include <sound/core.h>
+#include <sound/pcm.h>
+#include <sound/pcm_params.h>
+#include <sound/soc.h>
+#include <sound/tlv.h>
+
+#define AC200_CODEC_CAPTURE_RATES (SNDRV_PCM_RATE_8000 | \
+				   SNDRV_PCM_RATE_11025 | \
+				   SNDRV_PCM_RATE_12000 | \
+				   SNDRV_PCM_RATE_16000 | \
+				   SNDRV_PCM_RATE_22050 | \
+				   SNDRV_PCM_RATE_24000 | \
+				   SNDRV_PCM_RATE_32000 | \
+				   SNDRV_PCM_RATE_44100 | \
+				   SNDRV_PCM_RATE_48000)
+
+#define AC200_CODEC_PLAYBACK_RATES (AC200_CODEC_CAPTURE_RATES | \
+				    SNDRV_PCM_RATE_96000 | \
+				    SNDRV_PCM_RATE_192000)
+
+#define AC200_CODEC_FORMATS (SNDRV_PCM_FMTBIT_S8 | \
+			     SNDRV_PCM_FMTBIT_S16_LE | \
+			     SNDRV_PCM_FMTBIT_S20_LE | \
+			     SNDRV_PCM_FMTBIT_S24_LE | \
+			     SNDRV_PCM_FMTBIT_S32_LE)
+
+#define AC200_SYS_AUDIO_CTL0			0x0010
+#define AC200_SYS_AUDIO_CTL0_MCLK_GATING	BIT(1)
+#define AC200_SYS_AUDIO_CTL0_RST_INVALID	BIT(0)
+#define AC200_SYS_AUDIO_CTL1			0x0012
+#define AC200_SYS_AUDIO_CTL1_I2S_IO_EN		BIT(0)
+
+#define AC200_SYS_CLK_CTL			0x2000
+#define AC200_SYS_CLK_CTL_I2S			15
+#define AC200_SYS_CLK_CTL_ADC			3
+#define AC200_SYS_CLK_CTL_DAC			2
+#define AC200_SYS_MOD_RST			0x2002
+#define AC200_SYS_MOD_RST_I2S			15
+#define AC200_SYS_MOD_RST_ADC			3
+#define AC200_SYS_MOD_RST_DAC			2
+#define AC200_SYS_SR_CTL			0x2004
+#define AC200_SYS_SR_CTL_SR_MASK		GENMASK(3, 0)
+#define AC200_SYS_SR_CTL_SR(x)			(x)
+#define AC200_I2S_CTL				0x2100
+#define AC200_I2S_CTL_SDO_EN			3
+#define AC200_I2S_CTL_TX_EN			2
+#define AC200_I2S_CTL_RX_EN			1
+#define AC200_I2S_CTL_GEN			0
+#define AC200_I2S_CLK				0x2102
+#define AC200_I2S_CLK_BCLK_OUT			BIT(15)
+#define AC200_I2S_CLK_LRCK_OUT			BIT(14)
+#define AC200_I2S_CLK_BCLKDIV_MASK		GENMASK(13, 10)
+#define AC200_I2S_CLK_BCLKDIV(x)		((x) << 10)
+#define AC200_I2S_CLK_LRCK_MASK			GENMASK(9, 0)
+#define AC200_I2S_CLK_LRCK(x)			((x) - 1)
+#define AC200_I2S_FMT0				0x2104
+#define AC200_I2S_FMT0_MODE_MASK		GENMASK(15, 14)
+#define AC200_I2S_FMT0_MODE(x)			((x) << 14)
+#define AC200_I2S_FMT0_MODE_PCM			0
+#define AC200_I2S_FMT0_MODE_LEFT		1
+#define AC200_I2S_FMT0_MODE_RIGHT		2
+#define AC200_I2S_FMT0_TX_OFFSET_MASK		GENMASK(11, 10)
+#define AC200_I2S_FMT0_TX_OFFSET(x)		((x) << 10)
+#define AC200_I2S_FMT0_RX_OFFSET_MASK		GENMASK(9, 8)
+#define AC200_I2S_FMT0_RX_OFFSET(x)		((x) << 8)
+#define AC200_I2S_FMT0_SR_MASK			GENMASK(6, 4)
+#define AC200_I2S_FMT0_SR(x)			((x) << 4)
+#define AC200_I2S_FMT0_SW_MASK			GENMASK(3, 1)
+#define AC200_I2S_FMT0_SW(x)			((x) << 1)
+#define AC200_I2S_FMT1				0x2108
+#define AC200_I2S_FMT1_BCLK_POL_INVERT		BIT(15)
+#define AC200_I2S_FMT1_LRCK_POL_INVERT		BIT(14)
+#define AC200_I2S_MIX_SRC			0x2114
+#define AC200_I2S_MIX_SRC_LMIX_DAC		13
+#define AC200_I2S_MIX_SRC_LMIX_ADC		12
+#define AC200_I2S_MIX_SRC_RMIX_DAC		9
+#define AC200_I2S_MIX_SRC_RMIX_ADC		8
+#define AC200_I2S_MIX_GAIN			0x2116
+#define AC200_I2S_MIX_GAIN_LMIX_DAC		13
+#define AC200_I2S_MIX_GAIN_LMIX_ADC		12
+#define AC200_I2S_MIX_GAIN_RMIX_DAC		9
+#define AC200_I2S_MIX_GAIN_RMIX_ADC		8
+#define AC200_I2S_DAC_VOL			0x2118
+#define AC200_I2S_DAC_VOL_LEFT			8
+#define AC200_I2S_DAC_VOL_RIGHT			0
+#define AC200_I2S_ADC_VOL			0x211A
+#define AC200_I2S_ADC_VOL_LEFT			8
+#define AC200_I2S_ADC_VOL_RIGHT			0
+#define AC200_DAC_CTL				0x2200
+#define AC200_DAC_CTL_DAC_EN			15
+#define AC200_DAC_MIX_SRC			0x2202
+#define AC200_DAC_MIX_SRC_LMIX_DAC		13
+#define AC200_DAC_MIX_SRC_LMIX_ADC		12
+#define AC200_DAC_MIX_SRC_RMIX_DAC		9
+#define AC200_DAC_MIX_SRC_RMIX_ADC		8
+#define AC200_DAC_MIX_GAIN			0x2204
+#define AC200_DAC_MIX_GAIN_LMIX_DAC		13
+#define AC200_DAC_MIX_GAIN_LMIX_ADC		12
+#define AC200_DAC_MIX_GAIN_RMIX_DAC		9
+#define AC200_DAC_MIX_GAIN_RMIX_ADC		8
+#define AC200_OUT_MIX_CTL			0x2220
+#define AC200_OUT_MIX_CTL_RDAC_EN		15
+#define AC200_OUT_MIX_CTL_LDAC_EN		14
+#define AC200_OUT_MIX_CTL_RMIX_EN		13
+#define AC200_OUT_MIX_CTL_LMIX_EN		12
+#define AC200_OUT_MIX_CTL_MIC1_VOL		4
+#define AC200_OUT_MIX_CTL_MIC2_VOL		0
+#define AC200_OUT_MIX_SRC			0x2222
+#define AC200_OUT_MIX_SRC_RMIX_MIC1		14
+#define AC200_OUT_MIX_SRC_RMIX_MIC2		13
+#define AC200_OUT_MIX_SRC_RMIX_RDAC		9
+#define AC200_OUT_MIX_SRC_RMIX_LDAC		8
+#define AC200_OUT_MIX_SRC_LMIX_MIC1		6
+#define AC200_OUT_MIX_SRC_LMIX_MIC2		5
+#define AC200_OUT_MIX_SRC_LMIX_LDAC		1
+#define AC200_OUT_MIX_SRC_LMIX_RDAC		0
+#define AC200_LINEOUT_CTL			0x2224
+#define AC200_LINEOUT_CTL_EN			15
+#define AC200_LINEOUT_CTL_LEN			14
+#define AC200_LINEOUT_CTL_REN			13
+#define AC200_LINEOUT_CTL_LSRC			12
+#define AC200_LINEOUT_CTL_RSRC			11
+#define AC200_LINEOUT_CTL_SLOPE_MASK		GENMASK(9, 8)
+#define AC200_LINEOUT_CTL_SLOPE(x)		((x) << 8)
+#define AC200_LINEOUT_CTL_ANTIPOP_MASK		GENMASK(7, 5)
+#define AC200_LINEOUT_CTL_ANTIPOP(x)		((x) << 5)
+#define AC200_LINEOUT_CTL_VOL			0
+#define AC200_ADC_CTL				0x2300
+#define AC200_ADC_CTL_ADC_EN			15
+#define AC200_ADC_CTL_DELAY_TIME_MASK		GENMASK(3, 2)
+#define AC200_ADC_CTL_DELAY_TIME(x)		((x) << 2)
+#define AC200_ADC_CTL_DELAY_EN			BIT(1)
+#define AC200_MBIAS_CTL				0x2310
+#define AC200_MBIAS_CTL_MBIAS_EN		15
+#define AC200_MBIAS_CTL_VOLTAGE			8
+/*
+ * Undocumented: the datasheet lists MBIAS_CTRL[7:0] as unused and the vendor
+ * header does not name this bit either, but the ADC and DAC need it.
+ */
+#define AC200_MBIAS_CTL_ADDA_BIAS_EN		3
+#define AC200_ADC_MIC_CTL			0x2320
+#define AC200_ADC_MIC_CTL_RADC_EN		15
+#define AC200_ADC_MIC_CTL_LADC_EN		14
+#define AC200_ADC_MIC_CTL_ADC_VOL		8
+#define AC200_ADC_MIC_CTL_MIC1_GAIN_EN		7
+#define AC200_ADC_MIC_CTL_MIC1_BOOST		4
+#define AC200_ADC_MIC_CTL_MIC2_GAIN_EN		3
+#define AC200_ADC_MIC_CTL_MIC2_BOOST		0
+#define AC200_ADC_MIX_SRC			0x2322
+#define AC200_ADC_MIX_SRC_RMIX_MIC1		14
+#define AC200_ADC_MIX_SRC_RMIX_MIC2		13
+#define AC200_ADC_MIX_SRC_RMIX_RMIX		9
+#define AC200_ADC_MIX_SRC_RMIX_LMIX		8
+#define AC200_ADC_MIX_SRC_LMIX_MIC1		6
+#define AC200_ADC_MIX_SRC_LMIX_MIC2		5
+#define AC200_ADC_MIX_SRC_LMIX_LMIX		1
+#define AC200_ADC_MIX_SRC_LMIX_RMIX		0
+
+struct ac200_codec {
+	struct regmap *regmap;
+	unsigned int format;
+	unsigned int sysclk;
+};
+
+/*
+ * The AC200 expects a fixed MCLK of 24.576MHz or 22.5792MHz and divides it
+ * down according to SYS_SAMP_CTL, so the two sample rate families need
+ * different clocks and cannot be served at the same time.
+ */
+#define AC200_SYSCLK_48K	24576000
+#define AC200_SYSCLK_44K1	22579200
+
+static const unsigned int ac200_rates_48k[] = {
+	8000, 12000, 16000, 24000, 32000, 48000, 96000, 192000,
+};
+
+static const unsigned int ac200_rates_44k1[] = {
+	11025, 22050, 44100,
+};
+
+static const struct snd_pcm_hw_constraint_list ac200_constraints_48k = {
+	.list = ac200_rates_48k,
+	.count = ARRAY_SIZE(ac200_rates_48k),
+};
+
+static const struct snd_pcm_hw_constraint_list ac200_constraints_44k1 = {
+	.list = ac200_rates_44k1,
+	.count = ARRAY_SIZE(ac200_rates_44k1),
+};
+
+struct ac200_map {
+	int match;
+	int value;
+};
+
+static const SNDRV_CTL_TLVD_DECLARE_DB_SCALE(mixer_scale, -600, 600, 0);
+static const SNDRV_CTL_TLVD_DECLARE_DB_SCALE(gain_scale, -450, 150, 0);
+static const SNDRV_CTL_TLVD_DECLARE_DB_RANGE(lineout_scale,
+	0, 1, SNDRV_CTL_TLVD_DB_SCALE_ITEM(SNDRV_CTL_TLVD_DB_GAIN_MUTE, 0, 1),
+	2, 31, SNDRV_CTL_TLVD_DB_SCALE_ITEM(-4350, 150, 0),
+);
+static const SNDRV_CTL_TLVD_DECLARE_DB_SCALE(codec_scale, -12000, 75, 1);
+static const unsigned int mic_scale[] = {
+	TLV_DB_RANGE_HEAD(2),
+	0, 0, TLV_DB_SCALE_ITEM(0, 0, 0),
+	1, 7, TLV_DB_SCALE_ITEM(2400, 300, 0),
+};
+
+static const char * const mbias_voltage_enum_text[] = {
+	"1.88V", "2.09V", "2.33V", "2.50V",
+};
+
+static SOC_ENUM_SINGLE_DECL(mbias_voltage_enum, AC200_MBIAS_CTL,
+			    AC200_MBIAS_CTL_VOLTAGE, mbias_voltage_enum_text);
+
+static const struct snd_kcontrol_new ac200_codec_controls[] = {
+	SOC_DOUBLE_TLV("Master Playback Volume", AC200_I2S_DAC_VOL,
+		       AC200_I2S_DAC_VOL_LEFT, AC200_I2S_DAC_VOL_RIGHT,
+		       0xff, 0, codec_scale),
+	SOC_DOUBLE_TLV("Master Capture Volume", AC200_I2S_ADC_VOL,
+		       AC200_I2S_ADC_VOL_LEFT, AC200_I2S_ADC_VOL_RIGHT,
+		       0xff, 0, codec_scale),
+	SOC_DOUBLE_TLV("I2S ADC Capture Volume", AC200_I2S_MIX_GAIN,
+		       AC200_I2S_MIX_GAIN_LMIX_ADC, AC200_I2S_MIX_GAIN_RMIX_ADC,
+		       0x1, 1, mixer_scale),
+	SOC_DOUBLE_TLV("I2S DAC Capture Volume", AC200_I2S_MIX_GAIN,
+		       AC200_I2S_MIX_GAIN_LMIX_DAC, AC200_I2S_MIX_GAIN_RMIX_DAC,
+		       0x1, 1, mixer_scale),
+	SOC_DOUBLE_TLV("DAC I2S Playback Volume", AC200_DAC_MIX_GAIN,
+		       AC200_DAC_MIX_GAIN_LMIX_DAC, AC200_DAC_MIX_GAIN_RMIX_DAC,
+		       0x1, 1, mixer_scale),
+	SOC_DOUBLE_TLV("ADC Playback Volume", AC200_DAC_MIX_GAIN,
+		       AC200_DAC_MIX_GAIN_LMIX_ADC, AC200_DAC_MIX_GAIN_RMIX_ADC,
+		       0x1, 1, mixer_scale),
+	SOC_SINGLE_TLV("MIC1 Playback Volume", AC200_OUT_MIX_CTL,
+		       AC200_OUT_MIX_CTL_MIC1_VOL, 0x7, 0, gain_scale),
+	SOC_SINGLE_TLV("MIC2 Playback Volume", AC200_OUT_MIX_CTL,
+		       AC200_OUT_MIX_CTL_MIC2_VOL, 0x7, 0, gain_scale),
+	SOC_SINGLE_TLV("ADC Volume", AC200_ADC_MIC_CTL,
+		       AC200_ADC_MIC_CTL_ADC_VOL, 0x07, 0, gain_scale),
+	SOC_SINGLE_TLV("Line Out Playback Volume", AC200_LINEOUT_CTL,
+		       AC200_LINEOUT_CTL_VOL, 0x1f, 0, lineout_scale),
+	SOC_SINGLE_TLV("MIC1 Boost Volume", AC200_ADC_MIC_CTL,
+		       AC200_ADC_MIC_CTL_MIC1_BOOST, 0x07, 0, mic_scale),
+	SOC_SINGLE_TLV("MIC2 Boost Volume", AC200_ADC_MIC_CTL,
+		       AC200_ADC_MIC_CTL_MIC2_BOOST, 0x07, 0, mic_scale),
+	SOC_ENUM("Mic Bias Voltage", mbias_voltage_enum),
+};
+
+static const struct snd_kcontrol_new i2s_mixer[] = {
+	SOC_DAPM_DOUBLE("I2S DAC Capture Switch", AC200_I2S_MIX_SRC,
+			AC200_I2S_MIX_SRC_LMIX_DAC,
+			AC200_I2S_MIX_SRC_RMIX_DAC, 1, 0),
+	SOC_DAPM_DOUBLE("I2S ADC Capture Switch", AC200_I2S_MIX_SRC,
+			AC200_I2S_MIX_SRC_LMIX_ADC,
+			AC200_I2S_MIX_SRC_RMIX_ADC, 1, 0),
+};
+
+static const struct snd_kcontrol_new dac_mixer[] = {
+	SOC_DAPM_DOUBLE("DAC I2S Playback Switch", AC200_DAC_MIX_SRC,
+			AC200_DAC_MIX_SRC_LMIX_DAC,
+			AC200_DAC_MIX_SRC_RMIX_DAC, 1, 0),
+	SOC_DAPM_DOUBLE("ADC Playback Switch", AC200_DAC_MIX_SRC,
+			AC200_DAC_MIX_SRC_LMIX_ADC,
+			AC200_DAC_MIX_SRC_RMIX_ADC, 1, 0),
+};
+
+static const struct snd_kcontrol_new output_mixer[] = {
+	SOC_DAPM_DOUBLE("MIC1 Playback Switch", AC200_OUT_MIX_SRC,
+			AC200_OUT_MIX_SRC_LMIX_MIC1,
+			AC200_OUT_MIX_SRC_RMIX_MIC1, 1, 0),
+	SOC_DAPM_DOUBLE("MIC2 Playback Switch", AC200_OUT_MIX_SRC,
+			AC200_OUT_MIX_SRC_LMIX_MIC2,
+			AC200_OUT_MIX_SRC_RMIX_MIC2, 1, 0),
+	SOC_DAPM_DOUBLE("DAC Playback Switch", AC200_OUT_MIX_SRC,
+			AC200_OUT_MIX_SRC_LMIX_LDAC,
+			AC200_OUT_MIX_SRC_RMIX_RDAC, 1, 0),
+	SOC_DAPM_DOUBLE("DAC Reversed Playback Switch", AC200_OUT_MIX_SRC,
+			AC200_OUT_MIX_SRC_LMIX_RDAC,
+			AC200_OUT_MIX_SRC_RMIX_LDAC, 1, 0),
+};
+
+static const struct snd_kcontrol_new input_mixer[] = {
+	SOC_DAPM_DOUBLE("MIC1 Capture Switch", AC200_ADC_MIX_SRC,
+			AC200_ADC_MIX_SRC_LMIX_MIC1,
+			AC200_ADC_MIX_SRC_RMIX_MIC1, 1, 0),
+	SOC_DAPM_DOUBLE("MIC2 Capture Switch", AC200_ADC_MIX_SRC,
+			AC200_ADC_MIX_SRC_LMIX_MIC2,
+			AC200_ADC_MIX_SRC_RMIX_MIC2, 1, 0),
+	SOC_DAPM_DOUBLE("Output Mixer Capture Switch", AC200_ADC_MIX_SRC,
+			AC200_ADC_MIX_SRC_LMIX_LMIX,
+			AC200_ADC_MIX_SRC_RMIX_RMIX, 1, 0),
+	SOC_DAPM_DOUBLE("Output Mixer Reverse Capture Switch",
+			AC200_ADC_MIX_SRC,
+			AC200_ADC_MIX_SRC_LMIX_RMIX,
+			AC200_ADC_MIX_SRC_RMIX_LMIX, 1, 0),
+};
+
+/*
+ * The two source select bits are not symmetric: LSRC mixes the right
+ * output mixer into the left line out, while RSRC drives the right line
+ * out from the left one, to form a differential pair.
+ */
+static const char * const lineout_mux_enum_text[] = {
+	"Stereo", "Mono Differential",
+};
+
+static SOC_ENUM_DOUBLE_DECL(lineout_mux_enum, AC200_LINEOUT_CTL,
+			    AC200_LINEOUT_CTL_LSRC, AC200_LINEOUT_CTL_RSRC,
+			    lineout_mux_enum_text);
+
+static const struct snd_kcontrol_new lineout_mux[] = {
+	SOC_DAPM_ENUM("Line Out Source Playback Route", lineout_mux_enum),
+};
+
+static const struct snd_kcontrol_new lineout_switch =
+	SOC_DAPM_DOUBLE("Line Out Playback Switch", AC200_LINEOUT_CTL,
+			AC200_LINEOUT_CTL_LEN, AC200_LINEOUT_CTL_REN, 1, 0);
+
+static const struct snd_soc_dapm_widget ac200_codec_dapm_widgets[] = {
+	/*
+	 * DAPM only orders supplies against each other by subseq and then by
+	 * register address; supply to supply routes say whether a supply is
+	 * needed, not when it is switched. Everything below therefore carries
+	 * an explicit subseq:
+	 *
+	 *   0  module clocks       must precede the module resets
+	 *   1  module resets
+	 *   2  analog bias         must be settled before anything analog
+	 *   3  ADC/DAC digital
+	 *   4  I2S global enable   overrides the per direction enables
+	 *   5  I2S per direction
+	 */
+
+	/* System clocks */
+	SND_SOC_DAPM_SUPPLY_S("CLK SYS I2S", 0, AC200_SYS_CLK_CTL,
+			      AC200_SYS_CLK_CTL_I2S, 0, NULL, 0),
+	SND_SOC_DAPM_SUPPLY_S("CLK SYS DAC", 0, AC200_SYS_CLK_CTL,
+			      AC200_SYS_CLK_CTL_DAC, 0, NULL, 0),
+	SND_SOC_DAPM_SUPPLY_S("CLK SYS ADC", 0, AC200_SYS_CLK_CTL,
+			      AC200_SYS_CLK_CTL_ADC, 0, NULL, 0),
+
+	/* Module resets */
+	SND_SOC_DAPM_SUPPLY_S("RST SYS I2S", 1, AC200_SYS_MOD_RST,
+			      AC200_SYS_MOD_RST_I2S, 0, NULL, 0),
+	SND_SOC_DAPM_SUPPLY_S("RST SYS DAC", 1, AC200_SYS_MOD_RST,
+			      AC200_SYS_MOD_RST_DAC, 0, NULL, 0),
+	SND_SOC_DAPM_SUPPLY_S("RST SYS ADC", 1, AC200_SYS_MOD_RST,
+			      AC200_SYS_MOD_RST_ADC, 0, NULL, 0),
+
+	/* I2S gates */
+	SND_SOC_DAPM_SUPPLY_S("CLK I2S GEN", 4, AC200_I2S_CTL,
+			      AC200_I2S_CTL_GEN, 0, NULL, 0),
+	SND_SOC_DAPM_SUPPLY_S("CLK I2S SDO", 5, AC200_I2S_CTL,
+			      AC200_I2S_CTL_SDO_EN, 0, NULL, 0),
+	SND_SOC_DAPM_SUPPLY_S("CLK I2S TX", 5, AC200_I2S_CTL,
+			      AC200_I2S_CTL_TX_EN, 0, NULL, 0),
+	SND_SOC_DAPM_SUPPLY_S("CLK I2S RX", 5, AC200_I2S_CTL,
+			      AC200_I2S_CTL_RX_EN, 0, NULL, 0),
+
+	/* Module supplies */
+	SND_SOC_DAPM_SUPPLY_S("ADC Enable", 3, AC200_ADC_CTL,
+			      AC200_ADC_CTL_ADC_EN, 0, NULL, 0),
+	SND_SOC_DAPM_SUPPLY_S("DAC Enable", 3, AC200_DAC_CTL,
+			      AC200_DAC_CTL_DAC_EN, 0, NULL, 0),
+
+	/* Bias */
+	SND_SOC_DAPM_SUPPLY_S("MIC Bias", 2, AC200_MBIAS_CTL,
+			      AC200_MBIAS_CTL_MBIAS_EN, 0, NULL, 0),
+	SND_SOC_DAPM_SUPPLY_S("ADDA Bias", 2, AC200_MBIAS_CTL,
+			      AC200_MBIAS_CTL_ADDA_BIAS_EN, 0, NULL, 0),
+
+	/* DAC */
+	SND_SOC_DAPM_DAC("Left DAC", "Playback", AC200_OUT_MIX_CTL,
+			 AC200_OUT_MIX_CTL_LDAC_EN, 0),
+	SND_SOC_DAPM_DAC("Right DAC", "Playback", AC200_OUT_MIX_CTL,
+			 AC200_OUT_MIX_CTL_RDAC_EN, 0),
+
+	/* ADC */
+	SND_SOC_DAPM_ADC("Left ADC", "Capture", AC200_ADC_MIC_CTL,
+			 AC200_ADC_MIC_CTL_LADC_EN, 0),
+	SND_SOC_DAPM_ADC("Right ADC", "Capture", AC200_ADC_MIC_CTL,
+			 AC200_ADC_MIC_CTL_RADC_EN, 0),
+
+	/* Mixers */
+	SND_SOC_DAPM_MIXER("Left Output Mixer", AC200_OUT_MIX_CTL,
+			   AC200_OUT_MIX_CTL_LMIX_EN, 0,
+			   output_mixer, ARRAY_SIZE(output_mixer)),
+	SND_SOC_DAPM_MIXER("Right Output Mixer", AC200_OUT_MIX_CTL,
+			   AC200_OUT_MIX_CTL_RMIX_EN, 0,
+			   output_mixer, ARRAY_SIZE(output_mixer)),
+
+	SND_SOC_DAPM_MIXER("Left Input Mixer", SND_SOC_NOPM, 0, 0,
+			   input_mixer, ARRAY_SIZE(input_mixer)),
+	SND_SOC_DAPM_MIXER("Right Input Mixer", SND_SOC_NOPM, 0, 0,
+			   input_mixer, ARRAY_SIZE(input_mixer)),
+
+	SND_SOC_DAPM_MIXER("Left DAC Mixer", SND_SOC_NOPM, 0, 0,
+			   dac_mixer, ARRAY_SIZE(dac_mixer)),
+	SND_SOC_DAPM_MIXER("Right DAC Mixer", SND_SOC_NOPM, 0, 0,
+			   dac_mixer, ARRAY_SIZE(dac_mixer)),
+
+	SND_SOC_DAPM_MIXER("Left I2S Mixer", SND_SOC_NOPM, 0, 0,
+			   i2s_mixer, ARRAY_SIZE(i2s_mixer)),
+	SND_SOC_DAPM_MIXER("Right I2S Mixer", SND_SOC_NOPM, 0, 0,
+			   i2s_mixer, ARRAY_SIZE(i2s_mixer)),
+
+	/* Muxes */
+	SND_SOC_DAPM_MUX("Left Line Out Source", SND_SOC_NOPM, 0, 0,
+			 lineout_mux),
+	SND_SOC_DAPM_MUX("Right Line Out Source", SND_SOC_NOPM, 0, 0,
+			 lineout_mux),
+	SND_SOC_DAPM_SWITCH("Left Line Out Switch", SND_SOC_NOPM, 0, 0,
+			    &lineout_switch),
+	SND_SOC_DAPM_SWITCH("Right Line Out Switch", SND_SOC_NOPM, 0, 0,
+			    &lineout_switch),
+	SND_SOC_DAPM_OUT_DRV("Line Out Enable", AC200_LINEOUT_CTL,
+			     AC200_LINEOUT_CTL_EN, 0, NULL, 0),
+
+	/* Gain/attenuation */
+	SND_SOC_DAPM_PGA("MIC1 Amplifier", AC200_ADC_MIC_CTL,
+			 AC200_ADC_MIC_CTL_MIC1_GAIN_EN, 0, NULL, 0),
+	SND_SOC_DAPM_PGA("MIC2 Amplifier", AC200_ADC_MIC_CTL,
+			 AC200_ADC_MIC_CTL_MIC2_GAIN_EN, 0, NULL, 0),
+
+	/* Inputs */
+	SND_SOC_DAPM_INPUT("MIC1"),
+	SND_SOC_DAPM_INPUT("MIC2"),
+
+	/* Outputs */
+	SND_SOC_DAPM_OUTPUT("LINEOUT"),
+};
+
+static const struct snd_soc_dapm_route ac200_codec_dapm_routes[] = {
+	{ "RST SYS I2S", NULL, "CLK SYS I2S" },
+	{ "RST SYS ADC", NULL, "CLK SYS ADC" },
+	{ "RST SYS DAC", NULL, "CLK SYS DAC" },
+
+	{ "CLK I2S GEN", NULL, "RST SYS I2S" },
+	{ "CLK I2S SDO", NULL, "CLK I2S GEN" },
+	{ "CLK I2S TX", NULL, "CLK I2S SDO" },
+	{ "CLK I2S RX", NULL, "CLK I2S GEN" },
+
+	{ "ADC Enable", NULL, "RST SYS ADC" },
+	{ "ADC Enable", NULL, "ADDA Bias" },
+	{ "DAC Enable", NULL, "RST SYS DAC" },
+	{ "DAC Enable", NULL, "ADDA Bias" },
+
+	{ "Left DAC", NULL,  "DAC Enable" },
+	{ "Left DAC", NULL,  "CLK I2S RX" },
+	{ "Right DAC", NULL,  "DAC Enable" },
+	{ "Right DAC", NULL,  "CLK I2S RX" },
+
+	{ "Left ADC", NULL,  "ADC Enable" },
+	{ "Left ADC", NULL,  "CLK I2S TX" },
+	{ "Right ADC", NULL,  "ADC Enable" },
+	{ "Right ADC", NULL,  "CLK I2S TX" },
+
+	{ "Left Output Mixer", "MIC1 Playback Switch", "MIC1 Amplifier" },
+	{ "Left Output Mixer", "MIC2 Playback Switch", "MIC2 Amplifier" },
+	{ "Left Output Mixer", "DAC Playback Switch", "Left DAC Mixer" },
+	{ "Left Output Mixer", "DAC Reversed Playback Switch", "Right DAC Mixer" },
+
+	{ "Right Output Mixer", "MIC1 Playback Switch", "MIC1 Amplifier" },
+	{ "Right Output Mixer", "MIC2 Playback Switch", "MIC2 Amplifier" },
+	{ "Right Output Mixer", "DAC Playback Switch", "Right DAC Mixer" },
+	{ "Right Output Mixer", "DAC Reversed Playback Switch", "Left DAC Mixer" },
+
+	{ "Left Input Mixer", "MIC1 Capture Switch", "MIC1 Amplifier" },
+	{ "Left Input Mixer", "MIC2 Capture Switch", "MIC2 Amplifier" },
+	{ "Left Input Mixer", "Output Mixer Capture Switch", "Left Output Mixer" },
+	{ "Left Input Mixer", "Output Mixer Reverse Capture Switch", "Right Output Mixer" },
+
+	{ "Right Input Mixer", "MIC1 Capture Switch", "MIC1 Amplifier" },
+	{ "Right Input Mixer", "MIC2 Capture Switch", "MIC2 Amplifier" },
+	{ "Right Input Mixer", "Output Mixer Capture Switch", "Right Output Mixer" },
+	{ "Right Input Mixer", "Output Mixer Reverse Capture Switch", "Left Output Mixer" },
+
+	{ "Left I2S Mixer", "I2S DAC Capture Switch", "Left DAC" },
+	{ "Left I2S Mixer", "I2S ADC Capture Switch", "Left Input Mixer" },
+	{ "Right I2S Mixer", "I2S DAC Capture Switch", "Right DAC" },
+	{ "Right I2S Mixer", "I2S ADC Capture Switch", "Right Input Mixer" },
+
+	{ "Left DAC Mixer", "DAC I2S Playback Switch", "Left DAC" },
+	{ "Left DAC Mixer", "ADC Playback Switch", "Left Input Mixer" },
+	{ "Right DAC Mixer", "DAC I2S Playback Switch", "Right DAC" },
+	{ "Right DAC Mixer", "ADC Playback Switch", "Right Input Mixer" },
+
+	{ "Left Line Out Source", "Stereo", "Left Output Mixer" },
+	{ "Left Line Out Source", "Mono Differential", "Left Output Mixer" },
+	{ "Left Line Out Source", "Mono Differential", "Right Output Mixer" },
+	{ "Left Line Out Switch", "Line Out Playback Switch", "Left Line Out Source" },
+
+	{ "Right Line Out Switch", "Line Out Playback Switch", "Right Output Mixer" },
+	{ "Right Line Out Source", "Stereo", "Right Line Out Switch" },
+	{ "Right Line Out Source", "Mono Differential", "Left Line Out Switch" },
+
+	{ "Left ADC", NULL, "Left I2S Mixer" },
+	{ "Right ADC", NULL, "Right I2S Mixer" },
+
+	{ "Line Out Enable", NULL, "Left Line Out Switch" },
+	{ "Line Out Enable", NULL, "Right Line Out Source" },
+	{ "LINEOUT", NULL, "Line Out Enable" },
+
+	{ "MIC1", NULL, "MIC Bias" },
+	{ "MIC2", NULL, "MIC Bias" },
+	{ "MIC1 Amplifier", NULL, "MIC1" },
+	{ "MIC2 Amplifier", NULL, "MIC2" },
+};
+
+static int ac200_get_sr_sw(unsigned int width)
+{
+	switch (width) {
+	case 8:
+		return 1;
+	case 12:
+		return 2;
+	case 16:
+		return 3;
+	case 20:
+		return 4;
+	case 24:
+		return 5;
+	case 28:
+		return 6;
+	case 32:
+		return 7;
+	}
+
+	return -EINVAL;
+}
+
+static const struct ac200_map ac200_bclk_div_map[] = {
+	{ .match = 1,	.value = 1 },
+	{ .match = 2,	.value = 2 },
+	{ .match = 4,	.value = 3 },
+	{ .match = 6,	.value = 4 },
+	{ .match = 8,	.value = 5 },
+	{ .match = 12,	.value = 6 },
+	{ .match = 16,	.value = 7 },
+	{ .match = 24,	.value = 8 },
+	{ .match = 32,	.value = 9 },
+	{ .match = 48,	.value = 10 },
+	{ .match = 64,	.value = 11 },
+	{ .match = 96,	.value = 12 },
+	{ .match = 128,	.value = 13 },
+	{ .match = 176,	.value = 14 },
+	{ .match = 192,	.value = 15 },
+};
+
+static int ac200_get_bclk_div(unsigned int sysclk_rate,
+			      unsigned int sample_rate, unsigned int period)
+{
+	unsigned int div = sysclk_rate / sample_rate / period;
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(ac200_bclk_div_map); i++) {
+		const struct ac200_map *bdiv = &ac200_bclk_div_map[i];
+
+		if (bdiv->match == div)
+			return bdiv->value;
+	}
+
+	return -EINVAL;
+}
+
+static const struct ac200_map ac200_ssr_map[] = {
+	{ .match = 8000,	.value = 0 },
+	{ .match = 11025,	.value = 1 },
+	{ .match = 12000,	.value = 2 },
+	{ .match = 16000,	.value = 3 },
+	{ .match = 22050,	.value = 4 },
+	{ .match = 24000,	.value = 5 },
+	{ .match = 32000,	.value = 6 },
+	{ .match = 44100,	.value = 7 },
+	{ .match = 48000,	.value = 8 },
+	{ .match = 96000,	.value = 9 },
+	{ .match = 192000,	.value = 10 },
+};
+
+static int ac200_get_ssr(unsigned int sample_rate)
+{
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(ac200_ssr_map); i++) {
+		const struct ac200_map *ssr = &ac200_ssr_map[i];
+
+		if (ssr->match == sample_rate)
+			return ssr->value;
+	}
+
+	return -EINVAL;
+}
+
+static int ac200_codec_hw_params(struct snd_pcm_substream *substream,
+				 struct snd_pcm_hw_params *params,
+				 struct snd_soc_dai *dai)
+{
+	struct ac200_codec *priv = snd_soc_dai_get_drvdata(dai);
+	unsigned int slot_width = params_physical_width(params);
+	unsigned int sample_rate = params_rate(params);
+	unsigned int lrck_period, frame_width, sysclk;
+	int sr, sw, bclkdiv, ssr, ret;
+
+	sysclk = (sample_rate % 4000) ? AC200_SYSCLK_44K1 : AC200_SYSCLK_48K;
+	if (priv->sysclk && priv->sysclk != sysclk) {
+		dev_err(dai->dev,
+			"%u Hz needs a %u Hz MCLK, but MCLK is %u Hz\n",
+			sample_rate, sysclk, priv->sysclk);
+		return -EINVAL;
+	}
+
+	sr = ac200_get_sr_sw(params_width(params));
+	if (sr < 0) {
+		dev_err(dai->dev, "Unsupported sample width %u\n",
+			params_width(params));
+		return sr;
+	}
+
+	sw = ac200_get_sr_sw(slot_width);
+	if (sw < 0) {
+		dev_err(dai->dev, "Unsupported slot width %u\n", slot_width);
+		return sw;
+	}
+
+	switch (priv->format & SND_SOC_DAIFMT_FORMAT_MASK) {
+	case SND_SOC_DAIFMT_I2S:
+	case SND_SOC_DAIFMT_RIGHT_J:
+	case SND_SOC_DAIFMT_LEFT_J:
+		lrck_period = slot_width;
+		frame_width = slot_width * 2;
+		break;
+	case SND_SOC_DAIFMT_DSP_A:
+	case SND_SOC_DAIFMT_DSP_B:
+		lrck_period = slot_width * 2;
+		frame_width = lrck_period;
+		break;
+	default:
+		dev_err(dai->dev, "Unsupported DAI format 0x%x\n",
+			priv->format & SND_SOC_DAIFMT_FORMAT_MASK);
+		return -EINVAL;
+	}
+
+	bclkdiv = ac200_get_bclk_div(sysclk, sample_rate, frame_width);
+	if (bclkdiv < 0) {
+		dev_err(dai->dev, "No BCLK divider for %u Hz, %u bit frames\n",
+			sample_rate, frame_width);
+		return bclkdiv;
+	}
+
+	ssr = ac200_get_ssr(sample_rate);
+	if (ssr < 0) {
+		dev_err(dai->dev, "Unsupported sample rate %u\n", sample_rate);
+		return ssr;
+	}
+
+	ret = regmap_update_bits(priv->regmap, AC200_I2S_FMT0,
+				 AC200_I2S_FMT0_SR_MASK |
+				 AC200_I2S_FMT0_SW_MASK,
+				 AC200_I2S_FMT0_SR(sr) |
+				 AC200_I2S_FMT0_SW(sw));
+	if (ret)
+		return ret;
+
+	ret = regmap_update_bits(priv->regmap, AC200_I2S_CLK,
+				 AC200_I2S_CLK_LRCK_MASK |
+				 AC200_I2S_CLK_BCLKDIV_MASK,
+				 AC200_I2S_CLK_LRCK(lrck_period) |
+				 AC200_I2S_CLK_BCLKDIV(bclkdiv));
+	if (ret)
+		return ret;
+
+	return regmap_update_bits(priv->regmap, AC200_SYS_SR_CTL,
+				  AC200_SYS_SR_CTL_SR_MASK,
+				  AC200_SYS_SR_CTL_SR(ssr));
+}
+
+static int ac200_codec_startup(struct snd_pcm_substream *substream,
+			       struct snd_soc_dai *dai)
+{
+	struct ac200_codec *priv = snd_soc_dai_get_drvdata(dai);
+	const struct snd_pcm_hw_constraint_list *list;
+
+	switch (priv->sysclk) {
+	case AC200_SYSCLK_48K:
+		list = &ac200_constraints_48k;
+		break;
+	case AC200_SYSCLK_44K1:
+		list = &ac200_constraints_44k1;
+		break;
+	default:
+		/* MCLK is unknown, hw_params() will have to sort it out. */
+		return 0;
+	}
+
+	return snd_pcm_hw_constraint_list(substream->runtime, 0,
+					  SNDRV_PCM_HW_PARAM_RATE, list);
+}
+
+static int ac200_codec_set_sysclk(struct snd_soc_dai *dai, int clk_id,
+				  unsigned int freq, int dir)
+{
+	struct ac200_codec *priv = snd_soc_dai_get_drvdata(dai);
+
+	if (freq != AC200_SYSCLK_48K && freq != AC200_SYSCLK_44K1) {
+		dev_err(dai->dev, "Unsupported MCLK rate %u\n", freq);
+		return -EINVAL;
+	}
+
+	priv->sysclk = freq;
+
+	return 0;
+}
+
+static int ac200_codec_set_fmt(struct snd_soc_dai *dai, unsigned int fmt)
+{
+	struct ac200_codec *priv = snd_soc_dai_get_drvdata(dai);
+	unsigned int clk_value, inv_value, mode, offset;
+	int ret;
+
+	switch (fmt & SND_SOC_DAIFMT_CLOCK_PROVIDER_MASK) {
+	case SND_SOC_DAIFMT_CBP_CFP:
+		clk_value = AC200_I2S_CLK_BCLK_OUT | AC200_I2S_CLK_LRCK_OUT;
+		break;
+	case SND_SOC_DAIFMT_CBC_CFP:
+		clk_value = AC200_I2S_CLK_LRCK_OUT;
+		break;
+	case SND_SOC_DAIFMT_CBP_CFC:
+		clk_value = AC200_I2S_CLK_BCLK_OUT;
+		break;
+	case SND_SOC_DAIFMT_CBC_CFC:
+		clk_value = 0;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	switch (fmt & SND_SOC_DAIFMT_FORMAT_MASK) {
+	case SND_SOC_DAIFMT_I2S:
+		mode = AC200_I2S_FMT0_MODE_LEFT;
+		offset = 1;
+		break;
+	case SND_SOC_DAIFMT_RIGHT_J:
+		mode = AC200_I2S_FMT0_MODE_RIGHT;
+		offset = 0;
+		break;
+	case SND_SOC_DAIFMT_LEFT_J:
+		mode = AC200_I2S_FMT0_MODE_LEFT;
+		offset = 0;
+		break;
+	case SND_SOC_DAIFMT_DSP_A:
+		mode = AC200_I2S_FMT0_MODE_PCM;
+		offset = 1;
+		break;
+	case SND_SOC_DAIFMT_DSP_B:
+		mode = AC200_I2S_FMT0_MODE_PCM;
+		offset = 0;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	switch (fmt & SND_SOC_DAIFMT_INV_MASK) {
+	case SND_SOC_DAIFMT_NB_NF:
+		inv_value = 0;
+		break;
+	case SND_SOC_DAIFMT_NB_IF:
+		inv_value = AC200_I2S_FMT1_LRCK_POL_INVERT;
+		break;
+	case SND_SOC_DAIFMT_IB_NF:
+		inv_value = AC200_I2S_FMT1_BCLK_POL_INVERT;
+		break;
+	case SND_SOC_DAIFMT_IB_IF:
+		inv_value = AC200_I2S_FMT1_BCLK_POL_INVERT |
+			    AC200_I2S_FMT1_LRCK_POL_INVERT;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	ret = regmap_update_bits(priv->regmap, AC200_I2S_CLK,
+				 AC200_I2S_CLK_BCLK_OUT |
+				 AC200_I2S_CLK_LRCK_OUT, clk_value);
+	if (ret)
+		return ret;
+
+	ret = regmap_update_bits(priv->regmap, AC200_I2S_FMT0,
+				 AC200_I2S_FMT0_MODE_MASK |
+				 AC200_I2S_FMT0_TX_OFFSET_MASK |
+				 AC200_I2S_FMT0_RX_OFFSET_MASK,
+				 AC200_I2S_FMT0_MODE(mode) |
+				 AC200_I2S_FMT0_TX_OFFSET(offset) |
+				 AC200_I2S_FMT0_RX_OFFSET(offset));
+	if (ret)
+		return ret;
+
+	ret = regmap_update_bits(priv->regmap, AC200_I2S_FMT1,
+				 AC200_I2S_FMT1_BCLK_POL_INVERT |
+				 AC200_I2S_FMT1_LRCK_POL_INVERT,
+				 inv_value);
+	if (!ret)
+		priv->format = fmt;
+
+	return ret;
+}
+
+static const struct snd_soc_dai_ops ac200_codec_dai_ops = {
+	.startup	= ac200_codec_startup,
+	.hw_params	= ac200_codec_hw_params,
+	.set_sysclk	= ac200_codec_set_sysclk,
+	.set_fmt	= ac200_codec_set_fmt,
+};
+
+static struct snd_soc_dai_driver ac200_codec_dai = {
+	.name = "ac200-dai",
+	.playback = {
+		.stream_name = "Playback",
+		.channels_min = 2,
+		.channels_max = 2,
+		.rates = AC200_CODEC_PLAYBACK_RATES,
+		.formats = AC200_CODEC_FORMATS,
+	},
+	.capture = {
+		.stream_name = "Capture",
+		.channels_min = 2,
+		.channels_max = 2,
+		.rates = AC200_CODEC_CAPTURE_RATES,
+		.formats = AC200_CODEC_FORMATS,
+	},
+	.ops = &ac200_codec_dai_ops,
+	.symmetric_rate = 1,
+	.symmetric_channels = 1,
+	.symmetric_sample_bits = 1,
+};
+
+static int ac200_codec_component_probe(struct snd_soc_component *component)
+{
+	struct ac200_codec *priv = snd_soc_component_get_drvdata(component);
+
+	snd_soc_component_init_regmap(component, priv->regmap);
+
+	return 0;
+}
+
+static const struct snd_soc_component_driver ac200_soc_component = {
+	.controls		= ac200_codec_controls,
+	.num_controls		= ARRAY_SIZE(ac200_codec_controls),
+	.dapm_widgets		= ac200_codec_dapm_widgets,
+	.num_dapm_widgets	= ARRAY_SIZE(ac200_codec_dapm_widgets),
+	.dapm_routes		= ac200_codec_dapm_routes,
+	.num_dapm_routes	= ARRAY_SIZE(ac200_codec_dapm_routes),
+	.probe			= ac200_codec_component_probe,
+};
+
+static void ac200_codec_disable(void *data)
+{
+	struct ac200_codec *priv = data;
+
+	regmap_write(priv->regmap, AC200_SYS_AUDIO_CTL1, 0);
+	regmap_write(priv->regmap, AC200_SYS_AUDIO_CTL0, 0);
+}
+
+static int ac200_codec_probe(struct platform_device *pdev)
+{
+	struct device *dev = &pdev->dev;
+	struct ac200_codec *priv;
+	int ret;
+
+	priv = devm_kzalloc(dev, sizeof(*priv), GFP_KERNEL);
+	if (!priv)
+		return -ENOMEM;
+
+	priv->regmap = dev_get_regmap(dev->parent, NULL);
+	if (!priv->regmap)
+		return dev_err_probe(dev, -ENODEV, "Parent has no regmap\n");
+
+	platform_set_drvdata(pdev, priv);
+
+	ret = regmap_write(priv->regmap, AC200_SYS_AUDIO_CTL0,
+			   AC200_SYS_AUDIO_CTL0_RST_INVALID |
+			   AC200_SYS_AUDIO_CTL0_MCLK_GATING);
+	if (ret)
+		return ret;
+
+	ret = devm_add_action_or_reset(dev, ac200_codec_disable, priv);
+	if (ret)
+		return ret;
+
+	/*
+	 * Register the cleanup before touching the second control register, so
+	 * that a later error cannot leave the audio block's clock running.
+	 * Registering the component afterwards also ensures devres unwinding
+	 * unregisters it before gating the clock and asserting reset.
+	 */
+	ret = regmap_write(priv->regmap, AC200_SYS_AUDIO_CTL1,
+			   AC200_SYS_AUDIO_CTL1_I2S_IO_EN);
+	if (ret)
+		return ret;
+
+	/*
+	 * Match the vendor-selected 524ms anti-pop and slope times. Without
+	 * this the output stage snaps on and off and the result is audible.
+	 */
+	ret = regmap_update_bits(priv->regmap, AC200_LINEOUT_CTL,
+				 AC200_LINEOUT_CTL_SLOPE_MASK |
+				 AC200_LINEOUT_CTL_ANTIPOP_MASK,
+				 AC200_LINEOUT_CTL_SLOPE(3) |
+				 AC200_LINEOUT_CTL_ANTIPOP(3));
+	if (ret)
+		return ret;
+
+	/*
+	 * Hold ADC data back for 30ms after the ADC is enabled, so that the
+	 * settling of the analog front end does not end up in the stream.
+	 */
+	ret = regmap_update_bits(priv->regmap, AC200_ADC_CTL,
+				 AC200_ADC_CTL_DELAY_TIME_MASK |
+				 AC200_ADC_CTL_DELAY_EN,
+				 AC200_ADC_CTL_DELAY_TIME(3) |
+				 AC200_ADC_CTL_DELAY_EN);
+	if (ret)
+		return ret;
+
+	ret = devm_snd_soc_register_component(dev, &ac200_soc_component,
+					      &ac200_codec_dai, 1);
+	if (ret)
+		return dev_err_probe(dev, ret, "Failed to register codec\n");
+
+	return 0;
+}
+
+static const struct of_device_id ac200_codec_match[] = {
+	{ .compatible = "x-powers,ac200-codec" },
+	{ }
+};
+MODULE_DEVICE_TABLE(of, ac200_codec_match);
+
+static struct platform_driver ac200_codec_driver = {
+	.driver = {
+		.name = "ac200-codec",
+		.of_match_table = ac200_codec_match,
+	},
+	.probe = ac200_codec_probe,
+};
+module_platform_driver(ac200_codec_driver);
+
+MODULE_DESCRIPTION("X-Powers AC200 Codec Driver");
+MODULE_AUTHOR("Jernej Skrabec <jernej.skrabec@gmail.com>");
+MODULE_LICENSE("GPL");
-- 
2.43.0

